File: 1995-finding-pairs-with-a-certain-sum/1995-finding-pairs-with-a-certain-sum.java
class FindSumPairs {

    int[] nums1, nums2;
    Map<Integer, Integer> nums2Freq = new HashMap<>();

    public FindSumPairs(int[] nums1, int[] nums2) {
        this.nums1 = nums1;
        this.nums2 = nums2;
        for(int num: nums2){
            nums2Freq.put(num, nums2Freq.getOrDefault(num, 0) + 1);
        }
    }
    
    public void add(int index, int val) {
        int oldVal = nums2[index];
        nums2Freq.put(oldVal, nums2Freq.get(oldVal) - 1);
        if(nums2Freq.get(oldVal) == 0)
            nums2Freq.remove(oldVal);
        nums2[index] += val;
        nums2Freq.put(nums2[index], nums2Freq.getOrDefault(nums2[index], 0) + 1);
    }
    
    public int count(int tot) {
        int res = 0;
        
        for(int num : nums1){
            int temp = tot;
            if(nums2Freq.containsKey(temp- num)){
                res += nums2Freq.get(temp - num);
            }
        }
        return res;
    }
}

/**
 * Your FindSumPairs object will be instantiated and called as such:
 * FindSumPairs obj = new FindSumPairs(nums1, nums2);
 * obj.add(index,val);
 * int param_2 = obj.count(tot);
 */

File: 1995-finding-pairs-with-a-certain-sum/1995-finding-pairs-with-a-certain-sum.rs
use std::collections::HashMap;
struct FindSumPairs {
    nums1 :Vec<i32>,
    nums2 : Vec<i32>,
    nums2_freq : HashMap<i32, i32>
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl FindSumPairs {

    fn new(nums1: Vec<i32>, nums2: Vec<i32>) -> Self {
        let mut nums2_freq = HashMap::new();
        for &num in &nums2{
            *nums2_freq.entry(num).or_insert(0)+=1;
        }
        FindSumPairs {nums1, nums2, nums2_freq}
    }
    
    fn add(&mut self, index: i32, val: i32) {
        let old_val = self.nums2[index as usize];
        let new_val = old_val + val;

        if let Some(freq) = self.nums2_freq.get_mut(&old_val){
            *freq -= 1;
            if *freq == 0{
                self.nums2_freq.remove(&old_val);
            }
        }

        *self.nums2_freq.entry(new_val).or_insert(0) += 1;
        self.nums2[index as usize] = new_val;
    }
    
    fn count(&self, tot: i32) -> i32 {
        let mut res = 0;
        for &num in &self.nums1{
            let comp = tot - num;
            if let Some(&freq) = self.nums2_freq.get(&comp){
                res += freq;
            }
        }
        res
    }
}

/**
 * Your FindSumPairs object will be instantiated and called as such:
 * let obj = FindSumPairs::new(nums1, nums2);
 * obj.add(index, val);
 * let ret_2: i32 = obj.count(tot);
 */